### DISCUSSION TITLE
Issue with PATH suspected (can't get script to execute)

### DISCUSSION THREAD

**Discussion by benjaminwelch** [Taitava/obsidian-shellcommands#411] [Debugging and testing]
### Obsidian debug information

SYSTEM INFO:
	Obsidian version: v1.5.12
	Installer version: v1.4.16
	Operating system: Darwin Kernel Version 22.5.0: Mon Apr 24 20:53:19 PDT 2023; root:xnu-8796.121.2~5/RELEASE_ARM64_T6020 22.5.0
	Login status: not logged in
	Insider build toggle: off
	Live preview: on
	Base theme: light
	Community theme: Minimal v7.5.1
	Snippets enabled: 6
	Restricted mode: off
	Plugins installed: 40
	Plugins enabled: 38
		1: Calendar v1.5.10
		2: Dataview v0.5.66
		3: Electron Window Tweaker v1.0.8
		4: macOS Keyboard Navigation v1.1.2
		5: Minimal Theme Settings v7.5.0
		6: Regex Find/Replace v1.2.0
		7: Style Settings v1.0.8
		8: Tasks v7.1.0
		9: Timestamp Notes v1.0.8
		10: Buttons v0.5.1
		11: Advanced Tables v0.21.0
		12: BRAT v1.0.1
		13: Query Control v0.6.0
		14: Advanced URI v1.40.1
		15: Copy Block Link v1.0.4
		16: Kanban v1.5.3
		17: Media Extended v3.2.0
		18: Commander v0.5.1
		19: Paste URL into selection v1.7.0
		20: Outliner v4.8.1
		21: Templater v2.3.1
		22: List Callouts v1.2.7
		23: Natural Language Dates v0.6.2
		24: Iconize v2.12.0
		25: DB Folder v3.5.1
		26: Local REST API v2.3.0
		27: QuickAdd v1.8.1
		28: Remember File State v1.1.2
		29: Periodic Notes v0.0.17
		30: Note Refactor v1.8.2
		31: Metadata Menu v0.8.7
		32: Book Search v0.7.3
		33: Workspaces Plus v0.3.3
		34: Typewriter Scroll v0.2.2
		35: Tracker v1.14.0
		36: Breadcrumbs v3.6.11
		37: Creases v0.7.0
		38: Shell commands v0.22.0

RECOMMENDATIONS:
	Custom theme and snippets: for cosmetic issues, please first try updating your theme and disabling your snippets. If still not fixed, please try to make the issue happen in the Sandbox Vault or disable community theme and snippets.
	Community plugins: for bugs, please first try updating all your plugins to latest. If still not fixed, please try to make the issue happen in the Sandbox Vault or disable community plugins.

### What shell are you using?

Bash

### Please describe the problem

I'm trying to execute a Python script and when I do, I get this error:
```
Traceback (most recent call last):
  File "/Users/username/.system_scripts/things2md/things2md.py", line 14, in <module>
    from dateutil.relativedelta import *
ModuleNotFoundError: No module named 'dateutil'
```
I've installed `dateutil` for my global Python version and I can execute this Python script from the Terminal just fine. This leads me to believe that my shell command is not using the version of Python that I have enabled in my Terminal and this makes me think it's a PATH issue

Here's if what I've got:

## Shell command
`cd /Volumes/Macintosh\ HD/Users/username/.system_scripts/things2md/ && python3 things2md.py --today`

## Current PATH content:
`/usr/bin:/bin:/usr/sbin:/sbin`
I have no idea where this is being set/coming from.

## PATH from Terminal: echo $PATH
```
/Users/username/.pyenv/shims:/Users/username/miniconda3/condabin:/Users/username/Library/Python/3.9/bin:/Users/username/bin:/Users/username/.rbenv/shims:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Users/username/.local/bin
```
These aren't even close to the same and I'm not sure what to do about that.

I've also added this to my macOS PATH additions:
`/Volumes/Macintosh\ HD/Users/username/.pyenv/versions/3.12.1/bin/python`

Happy to provide more information and thanks for any help/suggestions.

Cheers,
Ben

**Comment by benjaminwelch** [comment]
I just thought of something else that might be relevant where PATHS are concerned. My Obsidian Vault is on an external SSD. I put the full path into my **macOS PATH additions** but I don't know if that will work if Obsidian isn't on the same drive.

If I try to use `pyenv which python` as a Shell command, it gives me this error: `/bin/bash: pyenv: command not found`

But it I just open my Terminal and type the same command, I see the current global version of Python.

Again, thanks for your help.

**Comment by Taitava** [comment]
Hi,

unfortunately the _Shell commands_ plugin invokes Bash (and Dash and Zsh, too) in a way that doesn't load Bash login scripts. This is probably the cause for why your `$PATH` looks very different.

You could try to fix this by [creating a custom shell configuration](https://publish.obsidian.md/shellcommands/Environments/Custom+shells/Settings+for+custom+shells#Configure+a+new+_custom+shell_):
- Shell name: `Bash with login scripts` (or something better).
- Executable binary path: `/bin/bash`
- Shell arguments:
    ```
    --login
    -c
    {{!shell_command_content}}
    ```
    This is the key part. [By using `--login` here, it makes Bash execute files `/etc/profile`, `~/.bash_profile`, `~/.bash_login`, and `~/.profile`](https://askubuntu.com/a/445015).
- Host operating system: macOS
- Shell's operating system: macOS
- Special characters escaping: Unix shell style
- Path translator: Can be left empty.
- Wrapper for shell command: Can be left empty.

Please let me know if this helps you or not? 🙂

**Reply by benjaminwelch** [reply]
Thanks @Taitava! I'll give this a try. Appreciate the help!
